File: benchmarks/c/array_ops.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define SIZE 1000
#define ITERATIONS 10000

/* Match blorp's Vector runtime: heap-allocated arrays with fresh
   allocation per vadd/scale operation (no in-place reuse). */

static long *vector_new(int len) {
    long *v = malloc(len * sizeof(long));
    return v;
}

static long *vadd_int(const long *a, const long *b, int len) {
    long *result = vector_new(len);
    for (int i = 0; i < len; i++) {
        result[i] = a[i] + b[i];
    }
    return result;
}

static long *scale_int(const long *v, long scalar, int len) {
    long *result = vector_new(len);
    for (int i = 0; i < len; i++) {
        result[i] = v[i] * scalar;
    }
    return result;
}

static long sum_int(const long *v, int len) {
    long sum = 0;
    for (int i = 0; i < len; i++) {
        sum += v[i];
    }
    return sum;
}

int main(void) {
    long *arr1 = vector_new(SIZE);
    long *arr2 = vector_new(SIZE);
    for (int i = 0; i < SIZE; i++) {
        arr1[i] = i;
        arr2[i] = i * 2;
    }

    long final_sum = 0;
    for (int iter = 0; iter < ITERATIONS; iter++) {
        long *combined = vadd_int(arr1, arr2, SIZE);
        long *scaled = scale_int(combined, 3, SIZE);
        final_sum += sum_int(scaled, SIZE);
        free(scaled);
        free(combined);
    }

    printf("Completed %d iterations, final sum: %ld\n", ITERATIONS, final_sum);
    free(arr1);
    free(arr2);
    return 0;
}

File: benchmarks/ocaml/array_ops.ml
let () =
  let size = 1000 in
  let iterations = 10_000 in
  let arr1 = Array.init size Fun.id in
  let arr2 = Array.init size (fun i -> i * 2) in
  let final_sum = ref 0 in
  for _iter = 0 to iterations - 1 do
    let combined = Array.make size 0 in
    for i = 0 to size - 1 do
      combined.(i) <- arr1.(i) + arr2.(i)
    done;
    let scaled = Array.make size 0 in
    for i = 0 to size - 1 do
      scaled.(i) <- combined.(i) * 3
    done;
    let sum = ref 0 in
    for i = 0 to size - 1 do
      sum := !sum + scaled.(i)
    done;
    final_sum := !final_sum + !sum
  done;
  Printf.printf "Completed %d iterations, final sum: %d\n" iterations !final_sum
